File: gateway-device-app/src/test/java/programmingtheiot/part03/integration/connection/CoapClientPerformanceTest.java
/**
 * 
 * This class is part of the Programming the Internet of Things
 * project, and is available via the MIT License, which can be
 * found in the LICENSE file at the top level of this repository.
 * 
 * Copyright (c) 2020 by Andrew D. King
 */ 

package programmingtheiot.part03.integration.connection;

import static org.junit.Assert.*;

import java.util.Set;
import java.util.logging.Logger;

import org.eclipse.californium.core.CoapClient;
import org.eclipse.californium.core.CoapResource;
import org.eclipse.californium.core.WebLink;
import org.junit.After;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Ignore;
import org.junit.Test;

import programmingtheiot.common.ConfigConst;
import programmingtheiot.common.ConfigUtil;
import programmingtheiot.common.DefaultDataMessageListener;
import programmingtheiot.common.IDataMessageListener;
import programmingtheiot.common.ResourceNameEnum;
import programmingtheiot.data.DataUtil;
import programmingtheiot.data.SensorData;
import programmingtheiot.data.SystemStateData;
import programmingtheiot.gda.connection.*;

/**
 * This test case class contains very basic integration tests for
 * CoapClientToServerConnectorTest. It should not be considered complete,
 * but serve as a starting point for the student implementing
 * additional functionality within their Programming the IoT
 * environment.
 *
 */
public class CoapClientPerformanceTest
{
	// NOTE: We'll use only 10,000 requests for CoAP
		public static final int MAX_TEST_RUNS = 10000;
		
		private CoapClientConnector coapClient = null;
		
		public static final int DEFAULT_TIMEOUT = 300 * 10000;
		public static final boolean USE_DEFAULT_RESOURCES = true;
		private static final Logger _Logger =
				Logger.getLogger(CoapClientConnectorTest.class.getName());

		@Before
		public void setUp() throws Exception
		{
			this.coapClient = new CoapClientConnector();
		}
		@Ignore
		@Test
		public void testPostRequestCon()
		{
			execTestPost(MAX_TEST_RUNS, true);
		}
		//@Ignore
		@Test
		public void testPostRequestNon()
		{
			execTestPost(MAX_TEST_RUNS, false);
		}
		
		@Ignore
		@Test
		public void testPutRequestCon()
		{
			execTestPut(MAX_TEST_RUNS, true);
		}
		@Ignore
		@Test
		public void testPutRequestNon()
		{
			execTestPut(MAX_TEST_RUNS, false);
		}
		
		private void execTestPost(int maxTestRuns, boolean enableCON)
		{
			SensorData sd = new SensorData();
			String payload = DataUtil.getInstance().sensorDataToJson(sd);
			
			long startMillis = System.currentTimeMillis();
			
			for (int seqNo = 0; seqNo < maxTestRuns; seqNo++) {
				this.coapClient.sendPostRequest(ResourceNameEnum.CDA_SENSOR_MSG_RESOURCE, enableCON, payload, DEFAULT_TIMEOUT);
			}
			
			long endMillis = System.currentTimeMillis();
			long elapsedMillis = endMillis - startMillis;
			
			_Logger.info("POST message - useCON " + enableCON + " [" + maxTestRuns + "]: " + elapsedMillis + " ms");
		}
		
		private void execTestPut(int maxTestRuns, boolean enableCON)
		{
			SensorData sd = new SensorData();
			String payload = DataUtil.getInstance().sensorDataToJson(sd);
			
			long startMillis = System.currentTimeMillis();
			
			for (int seqNo = 0; seqNo < maxTestRuns; seqNo++) {
				this.coapClient.sendPutRequest(ResourceNameEnum.CDA_SENSOR_MSG_RESOURCE, enableCON, payload, DEFAULT_TIMEOUT);
			}
			
			long endMillis = System.currentTimeMillis();
			long elapsedMillis = endMillis - startMillis;
			
			_Logger.info("PUT message - useCON " + enableCON + " [" + maxTestRuns + "]: " + elapsedMillis + " ms");
		}
}

File: constrained-device-app/src/test/python/programmingtheiot/part03/integration/connection/CoapClientPerformanceTest.py
#####
# 
# This class is part of the Programming the Internet of Things
# project, and is available via the MIT License, which can be
# found in the LICENSE file at the top level of this repository.
# 
# Copyright (c) 2020 by Andrew D. King
# 

import logging
import unittest

import time

from programmingtheiot.common.ResourceNameEnum import ResourceNameEnum

from programmingtheiot.cda.connection.CoapClientConnector import CoapClientConnector
from programmingtheiot.data.SensorData import SensorData
from programmingtheiot.data.DataUtil import DataUtil

class CoapClientPerformanceTest(unittest.TestCase):
	"""
	This test case class contains very basic integration tests for
	CoapClientConnector. It should not be considered complete,
	but serve as a starting point for the student implementing
	additional functionality within their Programming the IoT
	environment.
	"""
	
	NS_IN_MILLIS = 1000000
	
	# NOTE: We'll use only 10,000 requests for CoAP
	MAX_TEST_RUNS = 10000
	
	@classmethod
	def setUpClass(self):
		logging.disable(level = logging.WARNING)
	
	def setUp(self):
		self.coapClient = CoapClientConnector()

	def tearDown(self):
		pass
					
	@unittest.skip("Ignore for now.")
	def testPostRequestCon(self):
		print("Testing POST - CON")
		
		self._execTestPost(self.MAX_TEST_RUNS, True)

	#@unittest.skip("Ignore for now.")
	def testPostRequestNon(self):
		print("Testing POST - NON")
		
		self._execTestPost(self.MAX_TEST_RUNS, False)
		
	@unittest.skip("Ignore for now.")
	def testPutRequestCon(self):
		print("Testing PUT - CON")
		
		self._execTestPut(self.MAX_TEST_RUNS, True)

	@unittest.skip("Ignore for now.")
	def testPutRequestNon(self):
		print("Testing PUT - NON")
		
		self._execTestPut(self.MAX_TEST_RUNS, False)

	def _execTestPost(self, maxTestRuns: int, useCon: bool):
		sensorData = SensorData()
		payload = DataUtil().sensorDataToJson(sensorData)
		
		startTime = time.time_ns()
		
		for seqNo in range(0, maxTestRuns):
			
			self.coapClient.sendPostRequest(resource = ResourceNameEnum.CDA_SENSOR_MSG_RESOURCE, enableCON = useCon, payload = payload)
			
		endTime = time.time_ns()
		elapsedMillis = (endTime - startTime) / self.NS_IN_MILLIS
		
		print("POST message - useCON = " + str(useCon) + " [" + str(maxTestRuns) + "]: " + str(elapsedMillis) + " ms")
		
	def _execTestPut(self, maxTestRuns: int, useCon: bool):
		sensorData = SensorData()
		payload = DataUtil().sensorDataToJson(sensorData)
		
		startTime = time.time_ns()
		
		for seqNo in range(0, maxTestRuns):
			self.coapClient.sendPutRequest(resource = ResourceNameEnum.CDA_SENSOR_MSG_RESOURCE, enableCON = useCon, payload = payload)
			
		endTime = time.time_ns()
		elapsedMillis = (endTime - startTime) / self.NS_IN_MILLIS
		
		print("PUT message - useCON = " + str(useCon) + " [" + str(maxTestRuns) + "]: " + str(elapsedMillis) + " ms")

if __name__ == "__main__":
	unittest.main()
	